src/wfdirrd.c

```diff
@@ -1227,6 +1227,124 @@ DWORD DecodeReparsePoint(LPCWSTR szMyFile, LPCWSTR szChild, LPWSTR szDest, DWORD
+      else {
+         // If a Junction already exists, we have to check if it points to the 
+         // same location, and if yes then return ERROR_ALREADY_EXISTS
+         wchar_t szDestination[MAX_PATH] = { 0 };
+         AppendToPath(pszLinkDirectory, szStarDotStar);
+         DecodeReparsePoint(pszLinkDirectory, NULL, szDestination, COUNTOF(szDestination));
+         StripFilespec(pszLinkDirectory);
+         if (!_wcsicmp(szDestination, pszLinkTarget)) {
+            SetLastError(ERROR_ALREADY_EXISTS);
+            return ERROR_ALREADY_EXISTS;
+         }
+      }
+   }
+
+   HANDLE hFile = CreateFile(
+      pszLinkDirectory,
+      GENERIC_WRITE,
+      0,
+      NULL,
+      OPEN_EXISTING,
+      FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS,
+      NULL
+   );
+
+   if (INVALID_HANDLE_VALUE == hFile)
+      return GetLastError();
+
+   // Make the native target name
+   WCHAR szSubstituteName[MAX_PATH];
+
+   // The target might be
+   if (IsVeryLongPath(szTargetName)) {
+      // a very long path: \\?\x:\path\target
+      swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\%s", &szTargetName[PATH_PARSE_SWITCHOFF_SIZE]);
+   } else {
+      if (szTargetName[0] == L'\\' && szTargetName[1] == L'\\')
+         // an UNC name: \\myShare\path\target
+         swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\UNC\\%s", &szTargetName[2]);
+      else
+         // a normal full path: x:\path\target
+         swprintf_s(szSubstituteName, MAX_PATH, L"\\??\\%s", szTargetName);
+   }
+
+   // Delete the trailing slashes for non root path x:\path\foo\ -> x:\path\foo, but keep x:\
+   // Furthermore keep \\?\Volume{GUID}\ for 'root' volume-names
+   size_t lenSub = wcslen(szSubstituteName);
+   if ((szSubstituteName[lenSub - 1] == L'\\') && (szSubstituteName[lenSub - 2] != L':') && (szSubstituteName[lenSub - 2] != L'}'))
+      szSubstituteName[lenSub - 1] = 0;
+
+   PREPARSE_DATA_BUFFER reparseJunctionInfo = (PREPARSE_DATA_BUFFER)reparseBuffer;
+   memset(reparseJunctionInfo, 0, sizeof(REPARSE_DATA_BUFFER));
+   reparseJunctionInfo->ReparseTag = IO_REPARSE_TAG_MOUNT_POINT;
+
+   reparseJunctionInfo->MountPointReparseBuffer.SubstituteNameOffset = 0x00;
+   reparseJunctionInfo->MountPointReparseBuffer.SubstituteNameLength = (USHORT)(wcslen(szSubstituteName) * sizeof(wchar_t));
+   wcscpy_s(reparseJunctionInfo->MountPointReparseBuffer.PathBuffer, MAX_PATH, szSubstituteName);
+
+   reparseJunctionInfo->MountPointReparseBuffer.PrintNameOffset = reparseJunctionInfo->MountPointReparseBuffer.SubstituteNameLength + sizeof(wchar_t);
+   reparseJunctionInfo->MountPointReparseBuffer.PrintNameLength = (USHORT)(wcslen(szTargetName) * sizeof(wchar_t));
+   wcscpy_s(reparseJunctionInfo->MountPointReparseBuffer.PathBuffer + wcslen(szSubstituteName) + 1, MAX_PATH, szTargetName);
+
+   reparseJunctionInfo->ReparseDataLength = reparseJunctionInfo->MountPointReparseBuffer.SubstituteNameLength + reparseJunctionInfo->MountPointReparseBuffer.PrintNameLength + 12;

```

Comment:
Why 12?  Is this because 4 * sizeof(USHORT) + 2 * sizeof(wchar_t), where the USHORTs are in REPARSE_DATA_BUFFER and the wchar_t are the NULL terminators?

Fwiw in my version of this code I ended up with FIELD_OFFSET(REPARSE_DATA_BUFFER, MountPointReparseBuffer.PathBuffer) - FIELD_OFFSET(REPARSE_DATA_BUFFER, MountPointReparseBuffer) to express what the USHORTs are about.

Reply:
That's a good question. I took the coding a long time ago, I guess from a sample from Mark Russinovich.
It must be 12. Why 12? I don't know. Maybe you can look on the 'other side of the code', and find an explanation.

If I take your line instead of 12 it returns with ERROR_REPARSE_TAG_INVALID

I assume 12 is made up of  sizeof(ReparseTag) + sizeof (ReparseDataLength) + sizeof(Reserved) by using an alignment of 4. This turns into 12

But.... Only the Windows Source knows

Reply:
I used
`FIELD_OFFSET(REPARSE_DATA_BUFFER, MountPointReparseBuffer.PathBuffer[2]) - FIELD_OFFSET(REPARSE_DATA_BUFFER, MountPointReparseBuffer);`
and this resuls in 12